File: netlogo-gui/src/main/app/InterfacePanel.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.app;

import org.nlogo.api.I18N;
import org.nlogo.api.ModelSectionJ;
import org.nlogo.api.Version;
import org.nlogo.api.VersionHistory;
import org.nlogo.window.EditorColorizer;
import org.nlogo.window.Widget;

import javax.swing.JMenuItem;
import javax.swing.JPopupMenu;
import java.awt.event.KeyEvent;
import java.util.ArrayList;
import java.util.List;

strictfp class InterfacePanel
    extends WidgetPanel
    implements java.awt.event.KeyListener,
    org.nlogo.window.Events.LoadSectionEvent.Handler,
    org.nlogo.window.Events.ExportInterfaceEvent.Handler {

  private final org.nlogo.window.ViewWidgetInterface viewWidget;

  public org.nlogo.window.ViewWidgetInterface viewWidget() {
    return viewWidget;
  }

  public InterfacePanel(org.nlogo.window.ViewWidgetInterface viewWidget,
                        org.nlogo.window.GUIWorkspace workspace) {
    super(workspace);
    this.viewWidget = viewWidget;
    workspace.setWidgetContainer(this);
    // in 3d don't add the view widget since it's always
    // disabled there's no reason for it to take space 7/5/07
    if (!Version.is3D()) {
      addWidget((Widget) viewWidget, 0, 0, false, false);
    }
    ((Widget) viewWidget).deleteable_$eq(false);
    addKeyListener(this);
    addMouseListener(this);
  }

  ///

  @Override
  public void focusGained(java.awt.event.FocusEvent e) {
    hasFocus = true;
    enableButtonKeys(true);
  }

  @Override
  public void focusLost(java.awt.event.FocusEvent e) {
    hasFocus = false;
    enableButtonKeys(false);
  }

  ///

  @Override
  protected void doPopup(final java.awt.event.MouseEvent e) {
    JPopupMenu menu = new JPopupMenu();
    // add all the widgets
    menu.add(new WidgetCreationMenuItem(I18N.guiJ().get("tabs.run.widgets.button"), "BUTTON", e.getX(), e.getY()));
    menu.add(new WidgetCreationMenuItem(I18N.guiJ().get("tabs.run.widgets.slider"), "SLIDER", e.getX(), e.getY()));
    menu.add(new WidgetCreationMenuItem(I18N.guiJ().get("tabs.run.widgets.switch"), "SWITCH", e.getX(), e.getY()));
    menu.add(new WidgetCreationMenuItem(I18N.guiJ().get("tabs.run.widgets.chooser"), "CHOOSER", e.getX(), e.getY()));
    menu.add(new WidgetCreationMenuItem(I18N.guiJ().get("tabs.run.widgets.input"), "INPUT", e.getX(), e.getY()));
    menu.add(new WidgetCreationMenuItem(I18N.guiJ().get("tabs.run.widgets.monitor"), "MONITOR", e.getX(), e.getY()));
    menu.add(new WidgetCreationMenuItem(I18N.guiJ().get("tabs.run.widgets.plot"), "PLOT", e.getX(), e.getY()));
    WidgetCreationMenuItem outputItem =
        new WidgetCreationMenuItem(I18N.guiJ().get("tabs.run.widgets.output"), "OUTPUT", e.getX(), e.getY());
    if (getOutputWidget() != null) {
      outputItem.setEnabled(false);
    }
    menu.add(outputItem);
    menu.add(new WidgetCreationMenuItem(I18N.guiJ().get("tabs.run.widgets.note"), "NOTE", e.getX(), e.getY()));

    // add extra stuff
    menu.add(new javax.swing.JPopupMenu.Separator());
    menu.add(exportItem());

    menu.show(this, e.getX(), e.getY());
  }

  private JMenuItem exportItem() {
    JMenuItem exportItem = new javax.swing.JMenuItem("Export Interface...");
    exportItem.addActionListener
        (new java.awt.event.ActionListener() {
          public void actionPerformed(java.awt.event.ActionEvent e) {
            try {
              exportInterface();
            } catch (java.io.IOException ex) {
              javax.swing.JOptionPane.showMessageDialog
                  (InterfacePanel.this, ex.getMessage(),
                      I18N.guiJ().get("common.messages.error"), javax.swing.JOptionPane.ERROR_MESSAGE);
            }
          }
        });
    return exportItem;
  }

  private class WidgetCreationMenuItem
      extends javax.swing.JMenuItem {
    WidgetCreationMenuItem(final String displayName, final String widgetType, final int x, final int y) {
      super(displayName);
      addActionListener
          (new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent e) {
              Widget widget = makeWidget(widgetType, false);
              WidgetWrapper wrapper = addWidget(widget, x, y, true, false);
              revalidate();
              wrapper.selected(true);
              wrapper.foreground();
              wrapper.isNew(true);
              new org.nlogo.window.Events.EditWidgetEvent(null)
                  .raise(InterfacePanel.this);
              newWidget.setCursor
                  (java.awt.Cursor.getPredefinedCursor
                      (java.awt.Cursor.DEFAULT_CURSOR));
              wrapper.isNew(false);
              newWidget = null;
            }
          });
    }
  }

  // This is used both when loading a model and when the user is making
  // new widgets in the UI.  For most widget types, the same type string
  // is used in both places. - ST 3/17/04
  @Override
  public Widget makeWidget(String type, boolean loading) {
    type = type.toUpperCase();
    Widget fromRegistry = org.nlogo.window.WidgetRegistry.apply(type);
    if (fromRegistry != null) {
      return fromRegistry;
    } else if (type.equalsIgnoreCase("SLIDER")) {
      return new org.nlogo.window.SliderWidget(workspace.world.auxRNG) {
        @Override
        public int sourceOffset() {
          return org.nlogo.workspace.Evaluator.sourceOffset(org.nlogo.agent.Observer.class, false);
        }
      };
    } else if (type.equals("CHOOSER") || // current name
        type.equals("CHOICE"))   // old name, used in old models
    {
      return new org.nlogo.window.ChooserWidget(workspace);
    } else if (type.equals("BUTTON")) {
      return new org.nlogo.window.ButtonWidget(workspace.world.mainRNG);
    } else if (type.equals("PLOT")) {
      return org.nlogo.window.PlotWidget.apply(workspace.plotManager());
    } else if (type.equals("MONITOR")) {
      return new org.nlogo.window.MonitorWidget(workspace.world.auxRNG);
    } else if (type.equals("INPUT") ||  // in the GUI, it's "Input Box"
        type.equals("INPUTBOX"))  // in saved models, it's "INPUTBOX"
    {
      java.awt.Font font = new java.awt.Font(org.nlogo.awt.Fonts.platformMonospacedFont(),
          java.awt.Font.PLAIN, 12);
      return new org.nlogo.window.InputBoxWidget
          (new org.nlogo.window.CodeEditor
              (1, 20, font, false, null, new EditorColorizer(workspace), I18N.guiJ().fn()),
              new org.nlogo.window.CodeEditor
                  (5, 20, font, true, null, new EditorColorizer(workspace), I18N.guiJ().fn()),
              workspace, this);
    } else if (type.equals("OUTPUT"))  // currently in saved models only - ST 3/17/04
    {
      return new org.nlogo.window.OutputWidget();
    } else if (type.equals("CC-WINDOW")) // definitely in saved models only
    {
      // in current NetLogo versions, the command center goes in
      // a JSplitPane instead of in the InterfacePanel, so we ignore
      // the entry in the model - ST 7/13/04, 3/14/06
      return null;
    } else {
      throw new IllegalStateException
          ("unknown widget type: " + type);
    }
  }

  @Override
  protected void deleteWidgets(List<WidgetWrapper> hitList) {
    boolean needsRecompile = false;
    for (int i = 0; i < hitList.size(); i++) {
      WidgetWrapper w = hitList.get(i);
      removeWidget(w);
      if (w.widget() instanceof org.nlogo.window.JobWidget) {
        // this ensures that the right thing happens if we delete
        // a button or monitor that doesn't compile; we need to remove it
        // from the errors tab - ST 12/17/04
        org.nlogo.window.JobWidget jobWidget =
            (org.nlogo.window.JobWidget) w.widget();
        jobWidget.innerSource("");
        new org.nlogo.window.Events.CompileMoreSourceEvent(jobWidget)
            .raise(this);
      }
      if (w.widget() instanceof org.nlogo.window.InterfaceGlobalWidget) {
        needsRecompile = true;
      }
    }
    setForegroundWrapper();
    revalidate();
    repaint(); // you wouldn't think this'd be necessary, but without it
    // the widget didn't visually disappear - ST 6/23/03
    if (needsRecompile) {
      new org.nlogo.window.Events.CompileAllEvent().raise(this);
    }
    loseFocusIfAppropriate();
  }

  @Override
  protected void removeWidget(WidgetWrapper wrapper) {
    remove(wrapper);

    // if the compile that is associated with this removal (assuming there is one) fails
    // the observer variables and constraints might not get reallocated in which case
    // if we try to add a different widget with the same name we get a constraint violation
    // from the old constraint. yuck.  ev 11/27/07
    new org.nlogo.window.Events.RemoveConstraintEvent
        (wrapper.widget().displayName())
        .raise(this);

    org.nlogo.log.Logger.logWidgetRemoved(wrapper.widget().classDisplayName(),
        wrapper.widget().displayName());
  }

  /// loading and saving

  @Override
  public Widget loadWidget(String[] strings, final String modelVersion) {
    return loadWidget(strings, modelVersion, 0, 0);
  }

  // TODO: consider cleaning up this x and y business
  // it was added for copying/pasting widgets.
  // the regular loadWidget just uses the x and y from the string array
  // it passes in x=0, y=0 and we do a check. ugly, but works for now.
  // paste uses the x and y from the right click location.
  private Widget loadWidget(String[] strings, final String modelVersion, int x, int y) {
    Widget.LoadHelper helper =
        new Widget.LoadHelper() {
          public String version() {
            return modelVersion;
          }

          public String convert(String source, boolean reporter) {
            return workspace.autoConvert
                (source, true, reporter, modelVersion);
          }
        };
    String type = strings[0];
    if (x == 0) {
      x = Integer.parseInt(strings[1]);
    }
    if (y == 0) {
      y = Integer.parseInt(strings[2]);
    }
    if (viewWidget instanceof org.nlogo.window.ViewWidget &&
        !type.equals("GRAPHICS-WINDOW") &&
        VersionHistory.olderThan13pre1(modelVersion)) {
      y += ((org.nlogo.window.ViewWidget) viewWidget).getExtraHeight() +
          ((org.nlogo.window.ViewWidget) viewWidget).controlStrip.getHeight();
    }
    if (type.equals("GRAPHICS-WINDOW")) {
      // the graphics widget (and the command center) are special cases because
      // they are not recreated at load time, but reused
      viewWidget.asWidget().load(strings, helper);
      // in 3D we don't add the viewWidget to the interface panel
      // so don't worry about all the sizing junk ev 7/5/07
      java.awt.Container parent = viewWidget.asWidget().getParent();
      if (parent != null) {
        parent.setSize(viewWidget.asWidget().getSize());
        enforceMinimumAndMaximumWidgetSizes(viewWidget.asWidget());
        parent.setLocation(x, y);
        zoomer().zoomWidgetLocation
            (getWrapper(viewWidget.asWidget()),
                true, true, 1.0, zoomer().zoomFactor());
        zoomer().zoomWidgetSize
            (getWrapper(viewWidget.asWidget()),
                true, true, 1.0, zoomer().zoomFactor());
        zoomer().scaleComponentFont
            (((org.nlogo.window.ViewWidget) viewWidget)
                .view,
               zoomFactor(), 1.0, false);
      }
      return viewWidget.asWidget();
    } else {
      Widget newGuy = null;
      newGuy = makeWidget(type, true);
      if (newGuy != null) {
        newGuy.load(strings, helper);
        enforceMinimumAndMaximumWidgetSizes(newGuy);
        addWidget(newGuy, x, y, false, true);
      }
      return newGuy;
    }
  }

  @Override
  public List<org.nlogo.window.Widget> getWidgetsForSaving() {
    List<org.nlogo.window.Widget> result =
        new ArrayList<org.nlogo.window.Widget>();
    java.awt.Component[] comps = getComponents();
    // automatically add the view widget in 3D isn't not
    // in the comp list but we definitely want to save it
    // it won't be added twice as we're checking contains
    // below.  ev 7/5/07
    result.add((org.nlogo.window.Widget) viewWidget);
    // loop backwards so JLayeredPane gives us the components
    // in back-to-front order for saving - ST 9/29/03
    for (int i = comps.length - 1; i >= 0; i--) {
      if (comps[i] instanceof WidgetWrapper) {
        WidgetWrapper wrapper = (WidgetWrapper) comps[i];
        Widget widget = wrapper.widget();
        if (!result.contains(widget)) {
          result.add(widget);
        }
      }
    }
    return result;
  }

  @Override
  boolean contains(org.nlogo.api.Editable w) {
    if (w == ((Widget) viewWidget).getEditable()) {
      return true;
    }
    return super.contains(w);
  }

  @Override
  public void handle(org.nlogo.window.Events.WidgetRemovedEvent e) {
  }

  public void handle(org.nlogo.window.Events.ExportInterfaceEvent e) {
    try {
      javax.imageio.ImageIO.write
          (org.nlogo.awt.Images.paintToImage(this),
              "png", e.stream);
    } catch (java.io.IOException ex) {
      e.exceptionBox[0] = ex;
    }
  }

  private void exportInterface()
      throws java.io.IOException {
    try {
      final String exportPath =
          org.nlogo.swing.FileDialog.show
              (this, "Export Interface",
                  java.awt.FileDialog.SAVE,
                  workspace.guessExportName("interface.png"));
      final java.io.IOException[] exception =
          new java.io.IOException[]{null};
      org.nlogo.swing.ModalProgressTask.apply(
        org.nlogo.awt.Hierarchy.getFrame(this),
        "Exporting...",
        new Runnable() {
          public void run() {
            try {
              workspace.exportInterface(exportPath);
            } catch (java.io.IOException ex) {
              exception[0] = ex;
            }}});
      if (exception[0] != null) {
        throw exception[0];
      }
    } catch (org.nlogo.awt.UserCancelException ex) {
      org.nlogo.util.Exceptions.ignore(ex);
    }
  }

  public void handle(org.nlogo.window.Events.LoadSectionEvent e) {
    if (e.section == ModelSectionJ.WIDGETS()) {
      loadWidgets(e.lines, e.version);
    }
  }

  @Override
  public void removeAllWidgets() {
    try {
      java.awt.Component[] comps = getComponents();
      setVisible(false);
      for (int i = 0; comps.length != i; i++) {
        if (comps[i] instanceof WidgetWrapper) {
          WidgetWrapper wrapper = (WidgetWrapper) comps[i];
          if (wrapper.widget() != viewWidget) {
            removeWidget(wrapper);
          }
        }
      }
    } catch (RuntimeException ex) {
      org.nlogo.util.Exceptions.handle(ex);
    } finally {
      setVisible(false);
    }
  }

  /// buttons

  @Override
  public boolean isFocusable() {
    java.awt.Component[] comps = getComponents();
    for (int i = 0; i < comps.length; i++) {
      if (comps[i] instanceof WidgetWrapper) {
        Widget widget = ((WidgetWrapper) comps[i]).widget();
        if (widget instanceof org.nlogo.window.InputBoxWidget) {
          return true;
        }
        if (widget instanceof org.nlogo.window.ButtonWidget) {
          org.nlogo.window.ButtonWidget button =
              (org.nlogo.window.ButtonWidget) widget;
          if (button.actionKey() != '\0' &&
              button.actionKey() != ' ') {
            return true;
          }
        }
      }
    }
    return false;
  }

  private org.nlogo.window.ButtonWidget findActionButton(char key) {
    java.awt.Component[] comps = getComponents();
    for (int i = 0; i < comps.length; i++) {
      if (comps[i] instanceof WidgetWrapper) {
        Widget widget = ((WidgetWrapper) comps[i]).widget();
        if (widget instanceof org.nlogo.window.ButtonWidget) {
          org.nlogo.window.ButtonWidget button =
              (org.nlogo.window.ButtonWidget) widget;
          if (Character.toUpperCase(button.actionKey()) ==
              Character.toUpperCase(key)) {
            return button;
          }
        }
      }
    }
    return null;
  }

  private void enableButtonKeys(boolean enabled) {
    java.awt.Component[] comps = getComponents();
    for (int i = 0; i < comps.length; i++) {
      if (comps[i] instanceof WidgetWrapper) {
        Widget widget = ((WidgetWrapper) comps[i]).widget();
        if (widget instanceof org.nlogo.window.ButtonWidget) {
          org.nlogo.window.ButtonWidget button =
              (org.nlogo.window.ButtonWidget) widget;
          button.keyEnabled(enabled);
        }
      }
    }
  }

  public void keyTyped(java.awt.event.KeyEvent e) {
    if (e.getKeyChar() != java.awt.event.KeyEvent.CHAR_UNDEFINED &&
        !e.isActionKey() &&
        (e.getModifiers() & getToolkit().getMenuShortcutKeyMask()) == 0) {
      org.nlogo.window.ButtonWidget button =
          findActionButton(e.getKeyChar());
      if (button != null) {
        button.keyTriggered();
      }
    }
  }

  public void keyPressed(KeyEvent evt) {
  }

  public void keyReleased(KeyEvent evt) {
  }

  @Override
  public boolean canAddWidget(String widget) {
    return (!widget.equals("Output")) || (getOutputWidget() == null);
  }
}

File: netlogo-gui/src/main/app/InterfacePanel.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.app

import java.awt.{ Cursor, FileDialog => AwtFileDialog }
import java.awt.event.{ ActionListener, ActionEvent, KeyEvent, FocusEvent, MouseEvent }
import java.io.IOException
import java.util.{ ArrayList, List => JList }

import javax.imageio.ImageIO
import javax.swing.{ JMenuItem, JPopupMenu, JOptionPane }

import org.nlogo.window.{ ButtonWidget, ChooserWidget, CodeEditor,
  EditorColorizer, GUIWorkspace, InputBoxWidget, InterfaceGlobalWidget,
  JobWidget, MonitorWidget, OutputWidget, PlotWidget, SliderWidget,
  ViewWidget, ViewWidgetInterface, Widget, WidgetInfo, WidgetRegistry }
import org.nlogo.api.{ Editable, Exceptions, ModelSection, Version, VersionHistory }
import org.nlogo.awt.{ Fonts, Hierarchy, Images, UserCancelException }
import org.nlogo.core.{ AgentKind, I18N, View => CoreView, Widget => CoreWidget,
  Button => CoreButton, Chooser => CoreChooser, InputBox => CoreInputBox,
  Monitor => CoreMonitor, Output => CoreOutput, Plot => CorePlot, Slider => CoreSlider,
  Switch => CoreSwitch, TextBox => CoreTextBox }
import org.nlogo.log.Logger
import org.nlogo.swing.{ FileDialog => SwingFileDialog }
import org.nlogo.swing.ModalProgressTask
import org.nlogo.window.Events.{ CompileAllEvent, CompileMoreSourceEvent,
  EditWidgetEvent, LoadWidgetsEvent, RemoveConstraintEvent }
import org.nlogo.workspace.Evaluator

import scala.collection.JavaConverters._

class InterfacePanel(val viewWidget: ViewWidgetInterface, workspace: GUIWorkspace)
  extends WidgetPanel(workspace)
  with java.awt.event.KeyListener
  with LoadWidgetsEvent.Handler
  with org.nlogo.window.Events.ExportInterfaceEvent.Handler {

  workspace.setWidgetContainer(this)
  // in 3d don't add the view widget since it's always
  // disabled there's no reason for it to take space 7/5/07
  if (!Version.is3D)
    addWidget(viewWidget.asInstanceOf[Widget], 0, 0, false, false)

  viewWidget.asInstanceOf[Widget].deleteable = false
  addKeyListener(this)
  addMouseListener(this)

  ///

  override def focusGained(e: FocusEvent): Unit = {
    _hasFocus = true
    enableButtonKeys(true)
  }

  override def focusLost(e: FocusEvent): Unit = {
    _hasFocus = false
    enableButtonKeys(false)
  }

  ///

  override protected def doPopup(e: MouseEvent): Unit = {
    val menu = new JPopupMenu()
    Seq(WidgetInfo.button,
      WidgetInfo.slider,
      WidgetInfo.switch,
      WidgetInfo.chooser,
      WidgetInfo.input,
      WidgetInfo.monitor,
      WidgetInfo.plot)
    .map(i => i.displayName -> i.widgetThunk)
    .foreach {
        case (displayName, widgetThunk) =>
          menu.add(new WidgetCreationMenuItem(displayName, widgetThunk(), e.getX, e.getY))
    }

    // add all the widgets
    val outputItem =
        new WidgetCreationMenuItem(I18N.gui.get("tabs.run.widgets.output"),
          CoreOutput(0, 0, 0, 0, 11), e.getX, e.getY)
    if (getOutputWidget != null) {
      outputItem.setEnabled(false)
    }
    menu.add(outputItem)

    menu.add(new WidgetCreationMenuItem(I18N.gui.get("tabs.run.widgets.note"), CoreTextBox(None, fontSize = 11, color = 0), e.getX, e.getY))

    // add extra stuff
    menu.add(new JPopupMenu.Separator())
    menu.add(exportItem)

    menu.show(this, e.getX, e.getY)
  }

  val exportItem: JMenuItem = {
    val exportAction = new ActionListener() {
      def actionPerformed(e: ActionEvent): Unit = {
        try {
          exportInterface()
        } catch  {
          case ex: IOException =>
            JOptionPane.showMessageDialog(InterfacePanel.this, ex.getMessage(),
              I18N.gui.get("common.messages.error"), JOptionPane.ERROR_MESSAGE)
        }
      }
    }
    val exportItem = new JMenuItem(
      I18N.gui.get("menu.file.export.interface"))
    exportItem.addActionListener(exportAction)
    exportItem
  }

  class WidgetCreationMenuItem(val displayName: String, val coreWidget: CoreWidget, x: Int, y: Int)
  extends JMenuItem(displayName) with ActionListener {
    addActionListener(this)

    override def actionPerformed(e: ActionEvent): Unit = {
      val widget = makeWidget(coreWidget)
      val wrapper = addWidget(widget, x, y, true, false)
      revalidate()
      wrapper.selected(true)
      wrapper.foreground()
      wrapper.isNew(true)
      new EditWidgetEvent(null).raise(InterfacePanel.this)
      newWidget.setCursor(Cursor.getPredefinedCursor(Cursor.DEFAULT_CURSOR))
      wrapper.isNew(false)
      newWidget = null
    }
  }

  // This is used both when loading a model and when the user is making
  // new widgets in the UI.  For most widget types, the same type string
  // is used in both places. - ST 3/17/04
  override def makeWidget(coreWidget: CoreWidget): Widget = {
    val fromRegistry = WidgetRegistry(coreWidget.getClass.getSimpleName)
    if (fromRegistry != null)
      fromRegistry
    else coreWidget match {
      case c: CoreChooser  => new org.nlogo.window.ChooserWidget(workspace)
      case b: CoreButton   => new ButtonWidget(workspace.world.mainRNG)
      case p: CorePlot     => PlotWidget(workspace.plotManager)
      case m: CoreMonitor  => new MonitorWidget(workspace.world.auxRNG)
      case s: CoreSlider =>
        new SliderWidget(workspace.world.auxRNG) {
          override def sourceOffset: Int =
            Evaluator.sourceOffset(AgentKind.Observer, false)
        }
      case i: CoreInputBox =>
        val font = Fonts.monospacedFont
        val textArea = new CodeEditor(1, 20, font, true, null, new EditorColorizer(workspace), I18N.gui.fn)
        val dialogTextArea = new CodeEditor(5, 20, font, false, null, new EditorColorizer(workspace), I18N.gui.fn)
        new InputBoxWidget(textArea, dialogTextArea, workspace, this)
      case _ =>
        throw new IllegalStateException("unknown widget type: " + coreWidget.getClass.getName)
    }
  }

  override private[app] def deleteWidgets(hitList: Seq[WidgetWrapper]): Unit = {
    var needsRecompile: Boolean = false
    for (wrapper <- hitList) {
      removeWidget(wrapper)
      wrapper.widget match {
        // this ensures that the right thing happens if we delete
        // a button or monitor that doesn't compile; we need to remove it
        // from the errors tab - ST 12/17/04
        case jobWidget: JobWidget =>
          jobWidget.innerSource = ""
          new CompileMoreSourceEvent(jobWidget).raise(this)
        case _ =>
      }
      wrapper.widget match {
        case _: InterfaceGlobalWidget => needsRecompile = true
        case _ =>
      }
    }
    setForegroundWrapper()
    revalidate()
    repaint() // you wouldn't think this'd be necessary, but without it
    // the widget didn't visually disappear - ST 6/23/03
    if (needsRecompile) {
      new CompileAllEvent().raise(this)
    }
    loseFocusIfAppropriate()
  }

  override protected def removeWidget(wrapper: WidgetWrapper): Unit = {
    remove(wrapper)

    // if the compile that is associated with this removal (assuming there is one) fails
    // the observer variables and constraints might not get reallocated in which case
    // if we try to add a different widget with the same name we get a constraint violation
    // from the old constraint. yuck.  ev 11/27/07
    new RemoveConstraintEvent(wrapper.widget.displayName).raise(this)

    Logger.logWidgetRemoved(wrapper.widget.classDisplayName,
        wrapper.widget.displayName)
  }

  /// loading and saving

  override def loadWidget(coreWidget: CoreWidget): Widget =
    loadWidget(coreWidget, 0, 0)

  // TODO: consider cleaning up this x and y business
  // it was added for copying/pasting widgets.
  // the regular loadWidget just uses the x and y from the string array
  // it passes in x=0, y=0 and we do a check. ugly, but works for now.
  // paste uses the x and y from the right click location.
  private def loadWidget(coreWidget: CoreWidget, _x: Int, _y: Int): Widget = {
    val x = if (_x == 0) coreWidget.left else _x
    val y = if (_y == 0) coreWidget.top  else _y
    coreWidget match {
      case view: CoreView =>
        // the graphics widget (and the command center) are special cases because
        // they are not recreated at load time, but reused
        viewWidget.asInstanceOf[ViewWidget].load(view)
        // in 3D we don't add the viewWidget to the interface panel
        // so don't worry about all the sizing junk ev 7/5/07
        val parent = viewWidget.asWidget.getParent
        if (parent != null) {
          parent.setSize(viewWidget.asWidget.getSize)
          enforceMinimumAndMaximumWidgetSizes(viewWidget.asWidget)
          parent.setLocation(x, y)
          zoomer.zoomWidgetLocation(
            getWrapper(viewWidget.asWidget),
                  true, true, 1.0, zoomer.zoomFactor)
          zoomer.zoomWidgetSize(
            getWrapper(viewWidget.asWidget),
                  true, true, 1.0, zoomer.zoomFactor)
          zoomer.scaleComponentFont(
            viewWidget.asInstanceOf[ViewWidget].view,
                 zoomFactor, 1.0, false)
        }
        viewWidget.asWidget
      case _ =>
        makeAndLoadWidget(coreWidget, x, y)
    }
  }

  override def getWidgetsForSaving: Seq[CoreWidget] =
    // automatically add the view widget since it isn't in
    // the components list in 3D - ev 7/5/07
    (viewWidget.model +: getComponents.reverse.collect {
      case wrapper: WidgetWrapper => wrapper.widget.model
    }).distinct

  override private[app] def contains(w: Editable): Boolean =
    if (w == viewWidget.asInstanceOf[Widget].getEditable)
      true
    else
      super.contains(w)

  override def handle(e: org.nlogo.window.Events.WidgetRemovedEvent): Unit = {
  }

  def handle(e: org.nlogo.window.Events.ExportInterfaceEvent): Unit = {
    try ImageIO.write(Images.paintToImage(this), "png", e.stream)
    catch {
      case ex: IOException =>
        e.exceptionBox(0) = ex
    }
  }

  @throws(classOf[java.io.IOException])
  private def exportInterface(): Unit = {
    try {
      val exportPath = SwingFileDialog.show(this,
        I18N.gui.get("dialog.interface.export.file"),
        AwtFileDialog.SAVE,
        workspace.guessExportName("interface.png"))
      var exception = Option.empty[IOException]
      val runExport = new Runnable() {
        def run(): Unit = {
          try workspace.exportInterface(exportPath)
          catch {
            case ex: IOException => exception = Some(ex)
          }}}
      ModalProgressTask(Hierarchy.getFrame(this),
        I18N.gui.get("dialog.interface.export.task"), runExport)
      exception.foreach(e => throw e)
    } catch {
      case ex: UserCancelException => Exceptions.ignore(ex)
    }
  }

  def handle(e: LoadWidgetsEvent): Unit = {
    loadWidgets(e.widgets)
  }

  override def removeAllWidgets(): Unit = {
    try {
      val comps = getComponents()
      setVisible(false)
      for (component <- getComponents) {
        component match {
          case w: WidgetWrapper if w.widget != viewWidget =>
            removeWidget(w)
          case _ =>
        }
      }
    } catch {
      case ex: RuntimeException => Exceptions.handle(ex)
    } finally {
      setVisible(false)
    }
  }

  /// buttons

  override def isFocusable: Boolean =
    getComponents.collect {
      case w: WidgetWrapper => w.widget
    }.exists {
      case _: InputBoxWidget => true
      case b: ButtonWidget   =>
        b.actionKey != '\u0000' && b.actionKey != ' '
      case _ => false
    }

  private def findActionButton(key: Char): ButtonWidget = {
    import java.lang.Character.toUpperCase
    getComponents.collect {
      case w: WidgetWrapper => w.widget
    }.collect {
      case b: ButtonWidget if toUpperCase(b.actionKey) == toUpperCase(key) => b
    }.headOption.orNull
  }

  private def enableButtonKeys(enabled: Boolean): Unit =
    getComponents.collect {
      case w: WidgetWrapper => w.widget
    }.foreach {
      case b: ButtonWidget => b.keyEnabled(enabled)
      case _ =>
    }

  def keyTyped(e: KeyEvent): Unit = {
    if (e.getKeyChar() != KeyEvent.CHAR_UNDEFINED &&
      !e.isActionKey &&
    (e.getModifiers & getToolkit.getMenuShortcutKeyMask) == 0) {
      Option(findActionButton(e.getKeyChar)).foreach { button =>
        button.keyTriggered()
      }
    }
  }

  def keyPressed(evt: KeyEvent): Unit = { }

  def keyReleased(evt: KeyEvent): Unit = { }

  override def canAddWidget(widget: String): Boolean = {
    return (widget != "Output" || getOutputWidget == null)
  }
}
